File: blind75/twoPointers/125-validPalindrome.js
// 125. Valid Palindrome

// A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.
// Given a string s, return true if it is a palindrome, or false otherwise.

/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    let newS = s.toLowerCase();
    let len = newS.length;
    for (let i = 0, j = len -1; i <= j; i++, j--) {
        while (newS.charAt(i).charCodeAt(0) < 'a'.charCodeAt(0) || newS.charAt(i).charCodeAt(0) > 'z'.charCodeAt(0))
            i++;
        while (newS.charAt(j).charCodeAt(0) < 'a'.charCodeAt(0) || newS.charAt(j).charCodeAt(0) > 'z'.charCodeAt(0))
            j--;
        if (newS.charAt(i) !== newS.charAt(j))
            return false;
    } 
    return true;
};

let s = "A man, a plan, a canal: Panama";
let s1 = "race a car";
let s2 = " ";

console.log(isPalindrome(s));
console.log(isPalindrome(s1));
console.log(isPalindrome(s2));

File: specified-learning专项训练/slidingWindowAndTwoPointers/twoPointers/exercices/125-validPalindrome.py
# 125. Valid Palindrome
# A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.
# Given a string s, return true if it is a palindrome, or false otherwise.

class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        l = 0
        r = len(s) - 1
        # isalnum 判断是否为数字或者字母
        while l < r:
            if not ((ord(s[l]) >= ord('a') and ord(s[l]) <= ord('z')) or ((ord(s[l]) >= ord('0') and ord(s[l]) <= ord('9')))) :
                l = l + 1
                continue
            if not ((ord(s[r]) >= ord('a') and ord(s[r]) <= ord('z')) or ((ord(s[r]) >= ord('0') and ord(s[r]) <= ord('9')))) :
                r = r - 1
                continue
            if s[l] != s[r]: return False
            l = l + 1
            r = r - 1
        return True

s = "A man, a plan, a canal: Panama"
print(Solution().isPalindrome(s))
s = "race a car"
print(Solution().isPalindrome(s))
s = " "
print(Solution().isPalindrome(s))
s = "0P"
print(Solution().isPalindrome(s))
